kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md

```diff
@@ -1193,9 +1200,319 @@ Precompiled Contracts
+  // ---------------------------------------------------------------------------------------
+    rule isValidBLS12Scalar(X) => X >=Int 0 andBool X <Int (1 <<Int 256)
+
+    syntax PrecompiledOp ::= "BLS12G1ADD"
+ // -------------------------------------
+    rule <k> BLS12G1ADD => #end EVMC_SUCCESS ... </k>
+         <output>
+            _ => #bls12point(BLS12G1Add
+                    (   ( Bytes2Int(substrBytes(CallData(), 0, 64), BE, Unsigned)
+                        , Bytes2Int(substrBytes(CallData(), 64, 128), BE, Unsigned)
+                        )
+                    ,   ( Bytes2Int(substrBytes(CallData(), 128, 192), BE, Unsigned)
+                        , Bytes2Int(substrBytes(CallData(), 192, 256), BE, Unsigned)
+                        )
+                    ))
+         </output>
+      requires lengthBytes( CallData() ) ==Int 256
+        andBool bls12ValidForAdd
+            ( Bytes2Int(substrBytes(CallData(), 0, 64), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 64, 128), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 128, 192), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 192, 256), BE, Unsigned)
+            )
+
+    rule <k> BLS12G1ADD => #end EVMC_PRECOMPILE_FAILURE ... </k>
+      requires lengthBytes( CallData() ) =/=Int 256
+        orBool notBool bls12ValidForAdd
+            ( Bytes2Int(substrBytes(CallData(), 0, 64), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 64, 128), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 128, 192), BE, Unsigned)
+            , Bytes2Int(substrBytes(CallData(), 192, 256), BE, Unsigned)
+            )
+
+    syntax Bool ::= bls12ValidForAdd(Int, Int, Int, Int)  [function, total]
+ // -----------------------------------------------------------------------
+    rule bls12ValidForAdd(X0, Y0, X1, Y1)
+        => true
+            andBool isValidBLS12Coordinate(X0)
+            andBool isValidBLS12Coordinate(Y0)
+            andBool isValidBLS12Coordinate(X1)
+            andBool isValidBLS12Coordinate(Y1)
+            andBool BLS12G1OnCurve((X0, Y0))
+            andBool BLS12G1OnCurve((X1, Y1))
+
+    syntax PrecompiledOp ::= "BLS12G1MSM"
+ // -------------------------------------
+    // TODO: implement `bls12G1Msm` as a hook, using Pippenger's algorithm (blst_p1s_mult_pippenger)
+    // However, note that the implementation of `g1_lincomb_fast` has the
+    // following comment:
+    //
+    //  * @remark While this function is significantly faster than g1_lincomb_naive(), we refrain from
+    //  * using it in security-critical places (like verification) because the blst Pippenger code has not
+    //  * been audited. In those critical places, we prefer using g1_lincomb_naive() which is much simpler.
+    //
+    // https://github.com/ethereum/c-kzg-4844/blob/cc33b779cd3a227f51b35ce519a83cf91d81ccea/src/common/lincomb.c#L54-L56
```

Comment:
@dwightguth, could you please take a look at this and advise?

Reply:
Even if the version preferred by the client is relatively straightforward, we should still implement this as a hook using established libraries. We do not want to be responsible for the maintenance burden of ensuring a cryptographic function is secure, which is a more onerous burden than simply ensuring functional correctness as seen here.